2024/12/solve-1.ys

```diff
@@ -1,100 +1,72 @@
 #!/usr/bin/env ys-0
 
-defn main(& args):
-  terrain =: init(args.0 || "input")
-  regions =: terrain:regions
-  println:
-    total-price: terrain regions
-
-defn init(file="input"):
-  file:slurp:lines
+defn main(input='input'): !:say
+  terrain =: input:slurp:lines
+  regions =: regions(terrain)
+  total-price: terrain regions
 
 defn total-price(terrain regions):
-  sum:
-    map _ regions:
-      fn(r): price(terrain r)
-
-defn price(terrain region):
-  =>: region:len * perimeter(terrain region)
+  defn price(region):
+    region.# * perimeter(terrain region)
+  sum: regions.map(price)
 
 defn perimeter(terrain region):
-  sum:
-    map _ region:seq:
-      fn(yx):
-        len:
-          grep _ yx:nesw:
-            fn(it): region.get(it):not
-
-defn regions(terrain):
-  first:
-    reduce _ [[] hash-set()] terrain:coordinates:
-      fn([regions visited] yx):
-        if visited.get(yx):
-          vector: regions visited
-          else:
-            region =: terrain.region(terrain.at(yx) vector(yx))
-            vector: regions.conj(region) visited.union(region)
+  region.map(
+    \(nesw(_).remove(\(region.get(_))).#)
```

Comment:
`_` is `%1` in an anonymous fn `\(...)`
In Clojure you can't nest `#(...)` but apparently it's fine for `\(...)` because it compiles to `(fn [& []] ...)`.
I'd never tried nesting before :)